src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs

```diff
@@ -0,0 +1,206 @@
+        private readonly ISystemClock _systemClock;
+        private readonly ILog _log;
+        private readonly TradingReportFilter _tradingReportFilter;
+
+        private readonly IClientAccountClient _clientAccountClient;
+        
+        private readonly Backend.Contracts.IAccountsApi _accountsApi;
+        private readonly Backend.Contracts.IAccountHistoryApi _accountHistoryApi;
+        private readonly Backend.Contracts.ITradeMonitoringReadingApi _tradeMonitoringReadingApi;
+
+        public TradingReportService(
+            IEmailService emailService,
+            IConvertService convertService,
+            ISystemClock systemClock, 
+            ILog log,
+            TradingReportFilter tradingReportFilter,
+            
+            IClientAccountClient clientAccountClient,
+            
+            Backend.Contracts.IAccountsApi accountsApi,
+            Backend.Contracts.IAccountHistoryApi accountHistoryApi,
+            Backend.Contracts.ITradeMonitoringReadingApi tradeMonitoringReadingApi)
+        {
+            _emailService = emailService;
+            _convertService = convertService;
+            _systemClock = systemClock;
+            _log = log;
+            _tradingReportFilter = tradingReportFilter;
+
+            _clientAccountClient = clientAccountClient;
+            
+            _accountsApi = accountsApi;
+            _accountHistoryApi = accountHistoryApi;
+            _tradeMonitoringReadingApi = tradeMonitoringReadingApi;
+        }
+
+        private bool LegalEntityPredicate(string legalEntity)
+        {
+            return string.IsNullOrWhiteSpace(_tradingReportFilter.LegalEntityRegex)
+                   || (!string.IsNullOrWhiteSpace(legalEntity)
+                       && Regex.IsMatch(legalEntity, _tradingReportFilter.LegalEntityRegex, RegexOptions.IgnoreCase));
+        }
+
+        public async Task PerformReporting()
+        {
+            var from = _systemClock.UtcNow.Date.AddMonths(-1);
+            var to = _systemClock.UtcNow.Date;
+            var reportMonth = _systemClock.UtcNow.Date.AddMonths(-1).Month;
+
+            await _log.WriteInfoAsync(nameof(TradingReportService), nameof(PerformReporting),
+                $"Report invoked for {reportMonth} month, period from {from:s} to {to:s}");
+
+            //gather data concurrently, await & filter & convert
+            var pendingPositionsTask = _tradeMonitoringReadingApi.PendingOrders();
+            var accountsTask = _accountsApi.GetAllAccounts();
+
+            var accounts = (await accountsTask).Where(x => LegalEntityPredicate(x.LegalEntity))
+                .Select(x => _convertService.Convert<DataReaderAccountBackendContract, Account>(x))
+                .ToList();
+            var clientIds = accounts.Select(x => x.ClientId).Distinct().ToArray();
+            var accountIds = accounts.Select(x => x.Id).Distinct().ToHashSet();
+            var accountClients = accounts.ToDictionary(x => x.Id, x => x.ClientId);
```

Comment:
accountClients.Keys will contain the same data like accountIds.
accountClients.ContainKey can be used